### PR TITLE
Improve JavaScript API documentation

### PR DESCRIPTION
**Pull request by fcollonval**
## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/13984

## Code changes

Change some code visibility that should not have been public
Bump TypeDoc to the latest 0.24.1
Starts the configuration to link to external documentation. There are certainly holes as the check must be done manually...; but the configuration is in place see `typedoc.js` config file for those that want to contribute.

## User-facing changes

None

## Backwards-incompatible changes

Fix improper visibility of commands and menu helper fonction.

### CODE REVIEW

File: docs/source/extension/extension_dev.rst [hunk 1]
```diff
@@ -196,7 +195,7 @@ The extension package containing the theme plugin must include all static assets
 
 See the `JupyterLab Light Theme <https://github.com/jupyterlab/jupyterlab/tree/master/packages/theme-light-extension>`__ for an example.
 
-See the `TypeScript theme cookiecutter <https://github.com/jupyterlab/theme-cookiecutter>`__ for a quick start to developing a theme plugin.
+See the `TypeScript extension cookiecutter <https://github.com/jupyterlab/extension-cookiecutter-ts>`__ (choosing ``theme`` as ``kind`` ) for a quick start to developing a theme plugin.
```
Comment (krassowski) [inline docs/source/extension/extension_dev.rst:198]: Thanks! Probably also worth updating line 88 in `docs/source/extension/extension_migration.rst` which still refers to theme-cookiecutter:

> The create-theme script has been removed. If you want to create a new theme extension, you should use the [Theme Cookiecutter](https://github.com/jupyterlab/theme-cookiecutter) instead.

File: docs/source/extension/extension_migration.rst [hunk 2]
```diff
@@ -288,15 +292,18 @@ bumped their major version (following semver convention). We want to point out p
...
+   * Plugin ``@jupyterlab/statusbar-extension:line-col-status`` renamed to ``@jupyterlab/codemirror-extension:line-col-status``
+   * ``HoverBox`` component moved from ``@jupyterlab/apputils`` to ``@jupyterlab/ui-components``.
+   * Remove ``STATUSBAR_PLUGIN_ID`` from public API.
 - ``@jupyterlab/terminal`` from 3.x to 4.x
-  Xterm.js upgraded from 4.x to 5.x
-  ``IThemeObject.selection`` renamed to ``selectionBackground``
+   * Xterm.js upgraded from 4.x to 5.x
+   * ``IThemeObject.selection`` renamed to ``selectionBackground``
+- ``@jupyterlab/terminal-extension`` from 3.x to 4.x
+   Remove ``addCommands`` from public API
```
Comment (krassowski) [inline docs/source/extension/extension_migration.rst:306]:
```suggestion
   Removed ``addCommands`` from public API
```

File: docs/source/extension/extension_migration.rst [hunk 3]
```diff
@@ -242,9 +242,13 @@ bumped their major version (following semver convention). We want to point out p
 - ``@jupyterlab/filebrowser-extension`` from 3.x to 4.x
    Remove command ``filebrowser:create-main-launcher``. You can replace by ``launcher:create`` (same behavior)
    All launcher creation actions are moved to ``@jupyterlab/launcher-extension``.
+- ``@jupyterlab/imageviewer-extension`` from 3.x to 4.x
+   Remove ``addCommands`` from public API
 - ``@jupyterlab/mainmenu`` from 3.x to 4.x
    * ``IMainMenu.addMenu`` signature changed from ``addMenu(menu: Menu, options?: IMainMenu.IAddOptions): void``
      to ``addMenu(menu: Menu, update?: boolean, options?: IMainMenu.IAddOptions): void``
+   * Remove ``createEditMenu``, ``createFileMenu``, ``createKernelMenu``, ``createViewMenu``, ``createRunMenu``,
```
Comment (krassowski) [inline docs/source/extension/extension_migration.rst:250]:
```suggestion
   * Removed ``createEditMenu``, ``createFileMenu``, ``createKernelMenu``, ``createViewMenu``, ``createRunMenu``,
```

File: docs/source/extension/extension_migration.rst [hunk 4]
```diff
@@ -242,9 +242,13 @@ bumped their major version (following semver convention). We want to point out p
 - ``@jupyterlab/filebrowser-extension`` from 3.x to 4.x
    Remove command ``filebrowser:create-main-launcher``. You can replace by ``launcher:create`` (same behavior)
    All launcher creation actions are moved to ``@jupyterlab/launcher-extension``.
+- ``@jupyterlab/imageviewer-extension`` from 3.x to 4.x
+   Remove ``addCommands`` from public API
```
Comment (krassowski) [inline docs/source/extension/extension_migration.rst:246]:
```suggestion
   Removed ``addCommands`` from public API
```

File: docs/source/extension/extension_migration.rst [hunk 5]
```diff
@@ -288,15 +292,18 @@ bumped their major version (following semver convention). We want to point out p
...
-  Plugin ``@jupyterlab/statusbar-extension:running-sessions-status`` renamed to ``@jupyterlab/apputils-extension:running-sessions-status``
-  Plugin ``@jupyterlab/statusbar-extension:line-col-status`` renamed to ``@jupyterlab/codemirror-extension:line-col-status``
-  ``HoverBox`` component moved from ``@jupyterlab/apputils`` to ``@jupyterlab/ui-components``.
+   * Setting ``@jupyterlab/statusbar-extension:plugin . startMode`` moved to ``@jupyterlab/application-extension:shell . startMode``
+   * Plugin ``@jupyterlab/statusbar-extension:mode-switch`` renamed to ``@jupyterlab/application-extension:mode-switch``
+   * Plugin ``@jupyterlab/statusbar-extension:kernel-status`` renamed to ``@jupyterlab/apputils-extension:kernel-status``
+   * Plugin ``@jupyterlab/statusbar-extension:running-sessions-status`` renamed to ``@jupyterlab/apputils-extension:running-sessions-status``
+   * Plugin ``@jupyterlab/statusbar-extension:line-col-status`` renamed to ``@jupyterlab/codemirror-extension:line-col-status``
+   * ``HoverBox`` component moved from ``@jupyterlab/apputils`` to ``@jupyterlab/ui-components``.
+   * Remove ``STATUSBAR_PLUGIN_ID`` from public API.
```
Comment (krassowski) [inline docs/source/extension/extension_migration.rst:301]:
```suggestion
   * Removed ``STATUSBAR_PLUGIN_ID`` from public API.
```

File: docs/source/extension/extension_migration.rst [hunk 6]
```diff
@@ -323,17 +330,17 @@ bumped their major version (following semver convention). We want to point out p
         The renderer id must follow the convention ``<ISettingRegistry.IPlugin.id>.<propertyName>``. This is to
         ensure a custom renderer is not used for property with the same name but different schema.
 - ``@jupyterlab/translation`` from 3.x to 4.x
-   Renamed the method ``locale`` into the property ``languageCode`` in the ``NullTranslator``
+  Renamed the method ``locale`` into the property ``languageCode`` in the ``NullTranslator``
 - ``@jupyterlab/vdom`` and ``@jupyterlab/vdom-extension`` have been removed.
-   The underlying [vdom](https://github.com/nteract/vdom) Python package is unmaintained.
-   So it was decided to drop it from core packages.
+  The underlying [vdom](https://github.com/nteract/vdom) Python package is unmaintained.
```
Comment (krassowski) [inline docs/source/extension/extension_migration.rst:335]: That's a markdown link in rst document:

Screenshot from 2023-04-12 20-42-34

we can either use rst syntax or keep bare URL as there is some nice formatting for GitHub URLs in our docs:

```suggestion
  The underlying https://github.com/nteract/vdom Python package is unmaintained.
```

File: typedoc.js [hunk 7]
```diff
@@ -101,14 +75,157 @@ const entryPoints = packages
 
 const exclude =
   packages.flatMap(p => [`packages/${p}/test`]) +
-  ['packages/application-extension/src/index.tsx', 'examples/example.spec.ts'];
+  ['packages/application-extension'];
 
 module.exports = {
   entryPoints,
   exclude,
+  externalSymbolLinkMappings: {
```
Comment (krassowski) [inline typedoc.js:83]: That's amazing! For anyone interested to see how it looks like:

Screenshot from 2023-04-12 20-47-32

**Review by krassowski — COMMENTED** [review]
:rocket:

There is something odd with the right sidebar on main index page, this could be an upstream bug (maybe https://github.com/TypeStrong/typedoc/issues/2224?):

| Before | Current |
|---|---|
|  Screenshot from 2023-04-12 20-52-38 | Screenshot from 2023-04-12 20-50-29 |

In the old version the modules were on the right and there was no "on this page" section.

Probably fine to leave as-is, it's minor.

File: docs/source/extension/extension_dev.rst (hunk 1 above)
Comment (fcollonval) [reply docs/source/extension/extension_dev.rst:198]: Good catch - I'll do

File: docs/source/extension/extension_migration.rst (hunk 6 above)
Comment (fcollonval) [reply docs/source/extension/extension_migration.rst:335]: Replace by RST syntax

File: docs/source/extension/extension_migration.rst [hunk 8]
```diff
@@ -353,17 +361,17 @@ bumped their major version (following semver convention). We want to point out p
         The renderer id must follow the convention ``<ISettingRegistry.IPlugin.id>.<propertyName>``. This is to
         ensure a custom renderer is not used for property with the same name but different schema.
 - ``@jupyterlab/translation`` from 3.x to 4.x
-   Renamed the method ``locale`` into the property ``languageCode`` in the ``NullTranslator``
+  Renamed the method ``locale`` into the property ``languageCode`` in the ``NullTranslator``
 - ``@jupyterlab/vdom`` and ``@jupyterlab/vdom-extension`` have been removed.
-   The underlying [vdom](https://github.com/nteract/vdom) Python package is unmaintained.
-   So it was decided to drop it from core packages.
+  The underlying `vdom https://github.com/nteract/vdom`_ Python package is unmaintained.
```
Comment (fcollonval) [inline docs/source/extension/extension_migration.rst:366]:
```suggestion
  The underlying `vdom <https://github.com/nteract/vdom>`_ Python package is unmaintained.
```

**Pull request merged**